intelmq/bots/experts/sieve/expert.py

```diff
@@ -0,0 +1,229 @@
+            if self.process_conjunction(conj, event):
+                return True
+        return False
+
+    def process_conjunction(self, conj, event):
+        for cond in conj.cond:
+            if not self.process_condition(cond, event):
+                return False
+        return True
+
+    def process_condition(self, cond, event):
+        match = cond.match
+        if match.__class__.__name__ == 'ExistMatch':
+            return self.process_exist_match(match.key, match.op, event)
+        elif match.__class__.__name__ == 'StringMatch':
+            return self.process_string_match(match.key, match.op, match.value, event)
+        elif match.__class__.__name__ == 'NumericMatch':
+            return self.process_numeric_match(match.key, match.op, match.value, event)
+        elif match.__class__.__name__ == 'IpRangeMatch':
+            return self.process_ip_range_match(match.key, match.range, event)
+        elif match.__class__.__name__ == 'Expression':
+            return self.match_expression(match, event)
+        pass
+
+    def process_exist_match(self, key, op, event):
+        if op == ':exists':
+            return key in event
+        elif op == ':notexists':
+            return key not in event
+
+    def process_string_match(self, key, op, value, event):
+        if key not in event:
+            return op == '!=' or op == '!~'
+
+        if value.__class__.__name__ == 'SingleStringValue':
+            return self.process_string_operator(event[key], op, value.value)
+        elif value.__class__.__name__ == 'StringValueList':
+            for val in value.values:
+                if self.process_string_operator(event[key], op, val.value):
+                    return True
+            return False
+
+    def process_string_operator(self, lhs, op, rhs):
+        if op == '==':
+            return lhs == rhs
+        elif op == '!=':
+            return lhs != rhs
+        elif op == ':contains':
+            return lhs.find(rhs) >= 0
+        elif op == '=~':
+            return re.search(rhs, lhs) is not None
+        elif op == '!~':
+            return re.search(rhs, lhs) is None
+
+    def process_numeric_match(self, key, op, value, event):
+        if key not in event:
+            return False
+
+        if value.__class__.__name__ == 'SingleNumericValue':
+            return self.process_numeric_operator(event[key], op, value.value)
+        elif value.__class__.__name__ == 'NumericValueList':
+            for val in value.values:
+                if self.process_numeric_operator(event[key], op, val.value):
+                    return True
+            return False
+
+    def process_numeric_operator(self, lhs, op, rhs):
+        return eval(str(lhs) + op + str(rhs))  # TODO graceful error handling
+
+    def process_ip_range_match(self, key, ip_range, event):
+        if key not in event:
+            return False
+
+        try:
+            addr = ipaddress.ip_address(event[key])
+        except ValueError:
```

Comment:
Since the event type is not enforced (any key can be specified as the left hand side of the `<<` operator), the value is validated here. We did discuss whether or not to restrict this operator to keys with a type corresponding to an IP address (https://github.com/antoinet/intelmq/issues/11), and I just realized I didn't implement it the suggested way.